### PR TITLE
replace is_canonically_isomorphic by == and some Dicts by IdDicts

### PR DESCRIPTION
**Pull request by simonbrandhorst**

### CODE REVIEW

File: experimental/Schemes/AffineSchemes.jl [hunk 1]
```diff
@@ -415,7 +415,7 @@ function issubset(
     Y::AbsSpec{BRT, RT}
   ) where {BRT, RT<:MPolyQuoLocalizedRing{<:Any, <:Any, <:Any, <:Any, <:MPolyPowersOfElement}}
   R = base_ring(OO(X))
```
Comment (afkafkafk13) [inline experimental/Schemes/AffineSchemes.jl]: Why do you replace base_ring(OO(Y)) by ambient_ring(Y) for Y, but not for X? (here and possibly further occurrences)

Comment (simonbrandhorst) [reply experimental/Schemes/AffineSchemes.jl]: In the end they are equivalent (at least I expect so). But for the sake of consistency I also removed the rest.

**Comment by simonbrandhorst** [comment]
@HechtiDerLachs  When I rung the following code on this branch I get some complicated error that I am unable to debug. 
Could you please take a look?
``` 
@testset "SpecOpen_4" begin
  R, (x,y) = QQ["x", "y"]
  S, (t) = PolynomialRing(QQ, ["t"])
  t = t[1]

  A = Spec(R)
  B = Spec(S)
  f = x^2+x^3 -y^2
  C = subscheme(A, f)

  phi = maximal_extension(C, B, [x//y])
  psi = maximal_extension(B, C, [(1-t^2)//t^2, (1-t^2)//t^3])
  psi = restrict(psi, SpecOpen(B, [t*(1-t^2)]), domain(phi))
  phi_res = restrict(phi, domain(phi), domain(psi))
  #psi_res = restrict(psi, domain(psi), domain(phi))
  p = compose(psi, phi)
  q = compose(phi_res, psi)
  phi1 = restrict(phi,domain(phi),codomain(phi))
  @test phi == phi1
  @test restrict(p, domain(p), domain(p)) == identity_map(domain(p))
  @test restrict(q, domain(q), domain(q)) == identity_map(domain(q))
end

File: experimental/Schemes/SpecOpen.jl [hunk 2]
```diff
@@ -204,57 +204,48 @@ function intersect(
     check::Bool=true
   )
   X = ambient(U)
-  ambient_ring(X) === ambient_ring(Y) || error("Schemes can not be compared")
+  ambient_ring(U) === ambient_ring(Y) || error("schemes do not compare")
```
Comment (simonbrandhorst) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
  ambient_ring(U) === ambient_ring(Y) || error("schemes can not be compared")
```

File: experimental/Schemes/SpecOpen.jl [hunk 3]
```diff
@@ -204,57 +204,48 @@ function intersect(
...
   )
   X = ambient(U) 
-  is_canonically_isomorphic(X, ambient(V)) || error("ambient schemes do not coincide")
+  X == ambient(V) || error("ambient schemes do not coincide")
   return SpecOpen(X, [a*b for a in gens(U) for b in gens(V)])
 end
 
 function Base.union(U::T, V::T) where {T<:SpecOpen}
-  is_canonically_isomorphic(ambient(U), ambient(V)) || error("the two open sets do not lay in the same ambient scheme")
+  ambient(U) == ambient(V) || error("the two open sets do not lie in the same ambient scheme")
```
Comment (simonbrandhorst) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
  ambient(U) == ambient(V) || error("the two open sets are not contained in the same ambient scheme")
```

File: experimental/Schemes/AffineSchemes.jl [hunk 4]
```diff
@@ -429,21 +429,21 @@ function issubset(
...
   return issubset(X, Y) && issubset(Y, X)
 end
 
-function is_canonically_isomorphic(X::AbsSpec, Y::EmptyScheme)
+function ==(X::AbsSpec, Y::EmptyScheme)
   return issubset(X, Y)
 end
 
-is_canonically_isomorphic(X::EmptyScheme, Y::Spec) = is_canonically_isomorphic(Y, X)
+==(X::EmptyScheme, Y::Spec) = (Y == X)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/AffineSchemes.jl]:
```suggestion
==(X::EmptyScheme, Y::AbsSpec) = (Y == X)
```

File: experimental/Schemes/SpecOpen.jl [hunk 5]
```diff
@@ -204,57 +204,50 @@ function intersect(
     check::Bool=true
   )
   X = ambient(U)
-  ambient_ring(X) === ambient_ring(Y) || error("Schemes can not be compared")
+  ambient_ring(U) === ambient_ring(Y) || error("schemes can not be compared")
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl:207]:
```suggestion
  ambient_ring(X) === ambient_ring(Y) || error("schemes can not be compared")
```
Do we even have `ambient_ring` for `SpecOpen`s ?

File: experimental/Schemes/SpecOpen.jl [hunk 6]
```diff
@@ -204,57 +204,50 @@ function intersect(
...
 
 function Base.union(U::T, V::T) where {T<:SpecOpen}
-  is_canonically_isomorphic(ambient(U), ambient(V)) || error("the two open sets do not lay in the same ambient scheme")
+  ambient(U) == ambient(V) || error("the two open sets are not contained in the same ambient scheme")
   return SpecOpen(ambient(U), vcat(gens(U), gens(V)))
 end
 
 function issubset(
-    Y::Spec, 
+    Y::Spec,
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
    Y::AbsSpec,
```

File: experimental/Schemes/SpecOpen.jl [hunk 7]
```diff
@@ -204,57 +204,50 @@ function intersect(
...
   )
+  ambient_ring(Y) === ambient_ring(U) || return false
+  issubset(Y, ambient(U)) || return false
   return one(OO(Y)) in ideal(OO(Y), gens(U))
 end
 
+
 function issubset(
     U::SpecOpen,
     Y::Spec
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
    Y::AbsSpec
```

File: experimental/Schemes/SpecOpen.jl [hunk 8]
```diff
@@ -204,57 +204,50 @@ function intersect(
...
   ) 
-  for V in affine_patches(U)
-    issubset(V, Y) || return false
-  end
-  return true
+  return all(issubset(V, Y) for V in affine_patches(U))
 end
 
 function issubset(U::T, V::T) where {T<:SpecOpen}
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]:
```suggestion
function issubset(U::SpecOpen, V::SpecOpen)
```

File: experimental/Schemes/SpecOpen.jl [hunk 9]
```diff
@@ -264,24 +257,11 @@ function issubset(U::T, V::T) where {T<:SpecOpen}
   #return issubset(complement(intersect(V, ambient(U))), complement(U))
 end
 
-function is_canonically_isomorphic(U::T, V::T) where {T<:SpecOpen}
+function ==(U::SpecSubset, V::SpecSubset)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl:260]:
```suggestion
function ==(U::SpecOpen, V::SpecOpen)
```

File: experimental/Schemes/SpecOpen.jl [hunk 10]
```diff
@@ -264,24 +257,11 @@ function issubset(U::T, V::T) where {T<:SpecOpen}
...
+  ambient_ring(U) === ambient_ring(V) || return false
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl:261]: So it seems this is really defined? I wouldn't mind introducing it, anyway...

File: experimental/Schemes/SpecOpen.jl [hunk 11]
```diff
@@ -753,23 +733,23 @@ function SpecOpenMor(X::SpecType, d::RET, Y::SpecType, e::RET, f::Vector{RET}; c
   return SpecOpenMor(U, V, [SpecMor(U[1], Y, OO(U[1]).(f), check=check)], check=check)
 end
 
-function pullback(f::SpecOpenMor)
+function pullback(f::SpecOpenMor; check::Bool=true)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/SpecOpen.jl]: Why introduce a `check` parameter here? This is a getter for an object associated to an already existing `SpecOpenMor`. That means: The `SpecOpenMor` has already been created and if anything, the checks should have been made then.

File: experimental/Schemes/Types.jl [hunk 12]
```diff
@@ -265,6 +265,13 @@ the list ``f₁,…,fᵣ`` as the *generators* for ``U``.
   end
 end
 
+########################################################################
+# Common type fo subsets of affine space                               #
+########################################################################
+
+SpecSubset = Union{SpecOpen,AbsSpec,PrincipalOpenSubset}
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Types.jl:272]: Now I see that the previous note was not a typo.

**Comment by HechtiDerLachs** [comment]
> ```r
> A = Spec(R)
> ```

What is `R` in your code?

File: experimental/Schemes/SpecOpen.jl (hunk 11 above)
Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: It was for debugging, but didn't help.

**Comment by simonbrandhorst** [comment]
> > ```r
> > A = Spec(R)
> > ```
> 
> What is `R` in your code?

Edited the code snippet. It is just part of the test file.

File: experimental/Schemes/SpecOpen.jl (hunk 8 above)
Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl]: Okay, I guess the equality of base rings are checked by equality of the ambient rings already.

File: experimental/Schemes/SpecOpen.jl (hunk 5 above)
Comment (simonbrandhorst) [reply experimental/Schemes/SpecOpen.jl:207]: yes

**Comment by HechtiDerLachs** [comment]
> Edited the code snippet. It is just part of the test file.

Yes, I was already guessing this was the case. And hopefully solved the problem.

The point was: The code was checking on an `AbsSpec` `X` to appear in a list `l`. Now the check for that was suddenly done using what previously used to be `is_canonically_isomorphic` and thus returned the wrong result. We probably have to take care of other instances of this problem.

**Comment by simonbrandhorst** [comment]
Thank you. I hope the tests pass now at least.

**Comment by afkafkafk13** [comment]
Ci still complains (but at another location), because you are using names in the test "is_open_embedding" from the preceding test "issubset"  without redefining them.

File: experimental/Schemes/SpecOpen.jl (hunk 11 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/SpecOpen.jl]: Yes, after I took up on it, I already guessed that this was the purpose. But that issue is resolved now and we should remove it.

**Comment by simonbrandhorst** [comment]
Okay, I think this should be it.

**Comment by HechtiDerLachs** [comment]
> Okay, I think this should be it.

Haha! I'm not sure...

Anyway: I saw that you wrote a lot of tests for `is_closed_embedding` which involve a `SpecOpen`. These are marked as broken since we do not have these methods. For the time being, I do not think they are super important, because the `SpecOpen`s are more a technical backbone and not necessarily user facing. At least, I think, we should make up our minds whether we can prevent the user from having to deal with `SpecOpen`s outside the glueings. But I might be wrong with my perception of the situation, so please correct me.

**Comment by simonbrandhorst** [comment]
I am okay for these things to be not implemented. The tests just keep track of it. 
Well I could imagine the user to come in contact with these. How else can we describe opens which are not affine?

**Comment by afkafkafk13** [comment]
I see the broken tests as important information for covering what is there and what is not there at present. In the end, 

I expect SpecOpen not to be user-facing in the sense of the interactive user, but a user, who does his own programming (in the sense of scripting, not necessarily contributing), might well run into situations where the need for them arises from the theoretical context.

File: experimental/Schemes/Types.jl [hunk 13]
```diff
@@ -475,8 +475,8 @@ Glueing of two affine schemes ``X ↩ U ≅ V ↪ Y`` along open subsets
     ambient(domain(f)) === X || error("the domain of the glueing morphism is not an open subset of the first argument")
     ambient(codomain(f)) === Y || error("the codomain of the glueing morphism is not an open subset of the second argument")
     if check
-      (domain(f) === codomain(g) &&
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Types.jl]: Why revert these changes? I think `===` was just the right thing here! This is what caused the other bug that @simonbrandhorst asked be to look at above.

Comment (simonbrandhorst) [reply experimental/Schemes/Types.jl]: I was going through the code and made sure that everything that was previously 
`is_canonically_isomorphic` is now `==` and not `===`.  
I can see that to construct a glue map we may want `===` and not `==` . So you can revert again.

Comment (HechtiDerLachs) [reply experimental/Schemes/Types.jl]: Alright. So this has been a mistake previously. I will do it, once the tests have passed, so that we at least once have a result on test coverage.

Comment (simonbrandhorst) [reply experimental/Schemes/Types.jl]: The coverage seems to be okay. It slightly increased. I would defer any further coverage increases to future PRs.

File: experimental/Schemes/Types.jl [hunk 14]
```diff
@@ -468,8 +475,8 @@ Glueing of two affine schemes ``X ↩ U ≅ V ↪ Y`` along open subsets
     ambient(domain(f)) === X || error("the domain of the glueing morphism is not an open subset of the first argument")
     ambient(codomain(f)) === Y || error("the codomain of the glueing morphism is not an open subset of the second argument")
     if check
-      (is_canonically_isomorphic(domain(f), codomain(g)) && 
-       is_canonically_isomorphic(domain(g), codomain(f))) || error("maps can not be isomorphisms")
+      (domain(f) == codomain(g) &&
+      domain(g) ==  codomain(f)) || error("maps can not be isomorphisms")
```
Comment (simonbrandhorst) [inline experimental/Schemes/Types.jl]:
```suggestion
      (domain(f) === codomain(g) &&
      domain(g) ===  codomain(f)) || error("maps can not be isomorphisms")
```

File: experimental/Schemes/Types.jl (hunk 13 above)
Comment (simonbrandhorst) [reply experimental/Schemes/Types.jl]: O.k. I reverted it. Let's see what the tests say.

**Review by HechtiDerLachs — APPROVED** [review]
